src/pkcs15init/pkcs15-jacartapki.c

```diff
@@ -0,0 +1,2098 @@
+	unsigned char *cmap = NULL;
+	size_t cmap_len, data_len;
+	int rv;
+
+	LOG_FUNC_CALLED(ctx);
+	sc_log(ctx, "Update CMAP; remove %i; object type 0x%X", remove, object ? object->type : (unsigned)(-1));
+	if (object != NULL && (object->type & SC_PKCS15_TYPE_CLASS_MASK) == SC_PKCS15_TYPE_PRKEY) {
+		info = (struct sc_pkcs15_prkey_info *)object->data;
+
+		if (info->aux_data != NULL && (info->aux_data->type != SC_AUX_DATA_TYPE_MD_CMAP_RECORD && info->aux_data->type != SC_AUX_DATA_TYPE_NO_DATA))
+			sc_aux_data_free(&info->aux_data);
+		if (info->aux_data == NULL) {
+			rv = sc_aux_data_allocate(ctx, &info->aux_data, NULL);
+			LOG_TEST_RET(ctx, rv, "Cannot allocate MD auxiliary data");
+		}
+
+		if (info->aux_data->type != SC_AUX_DATA_TYPE_MD_CMAP_RECORD || info->aux_data->data.cmap_record.guid_len == 0) {
+			int is_converted = 0;
+
+			rv = jacartapki_cmap_set_key_guid(ctx, info, &is_converted);
+			LOG_TEST_RET(ctx, rv, "Cannot set JaCarta PKI style GUID");
+		}
+
+		/* All new keys are 'key-exchange' keys.
+		 * FIXME: implement 'sign' key. */
+		info->aux_data->data.cmap_record.keysize_keyexchange = (unsigned int)info->modulus_length;
+		info->aux_data->data.cmap_record.keysize_sign = 0;
+
+		info->aux_data->data.cmap_record.flags = SC_MD_CONTAINER_MAP_VALID_CONTAINER;
+		sc_log(ctx, "Set 'valid container' flag for key object '%s'", sc_pkcs15_print_id(&info->id));
+	}
+
+	rv = jacartapki_cmap_container_set_default(p15card, remove, object);
+	LOG_TEST_RET(ctx, rv, "Failed to set default CMAP container");
+
+	rv = jacartapki_cmap_encode(p15card, (remove ? object : NULL), &cmap, &cmap_len);
+	LOG_TEST_RET(ctx, rv, "Failed to encode 'cmap' data");
+	sc_log(ctx, "encoded CMAP(%" SC_FORMAT_LEN_SIZE_T "u) '%s'", cmap_len, sc_dump_hex(cmap, cmap_len));
+
+	rv = sc_pkcs15_find_data_object_by_name(p15card, CMAP_DO_APPLICATION_NAME, "cmapfile", &cmap_dobj);
+	LOG_TEST_RET(ctx, rv, "Failed to get 'cmapfile' DATA object");
+
+	cmap_dobj_info = (struct sc_pkcs15_data_info *)cmap_dobj->data;
+
+	free(cmap_dobj_info->data.value);
+
+	data_len = cmap_len + sizeof(struct jacartapki_cmap_record);
+	if (data_len < 5 * sizeof(struct jacartapki_cmap_record))
+		data_len = 5 * sizeof(struct jacartapki_cmap_record);
+
+	cmap_dobj_info->data.value = calloc(1, data_len);
+	if (cmap_dobj_info->data.value == NULL) {
+		rv = SC_ERROR_OUT_OF_MEMORY;
+		LOG_ERROR_GOTO(ctx, rv, "Failed to allocate cmap object info");
+	}
+
+	memcpy(cmap_dobj_info->data.value, cmap, cmap_len);
+	cmap_dobj_info->data.len = data_len;
+
+	rv = jacartapki_update_df_create_data_object(profile, p15card, cmap_dobj);
+	LOG_TEST_GOTO_ERR(ctx, rv, "Failed to update DATA-DF ");
+err:
+	free(cmap);
+	LOG_FUNC_RETURN(ctx, rv);
+}
+
+static int
+jacartapki_cardcf_create(struct sc_profile *profile, struct sc_pkcs15_card *p15card, struct sc_file *file)
+{
+	struct sc_context *ctx = p15card->card->ctx;
+	struct jacartapki_cardcf cardcf = {
+			{0x00, 0x06, 0x00, 0x03},
+			0x1,
+			0x1
+	   };
```

Comment:
You need to add __attribute__((packed)) to struct laser_cardcf to make sure the size in bytes is correct